src/Core/Framework/MessageQueue/ScheduledTask/Scheduler/TaskScheduler.php

```diff
@@ -45,22 +45,19 @@ public function queueScheduledTasks(): void
             return;
         }
 
-        $updatePayload = [];
-        /** @var ScheduledTaskEntity $task */
-        foreach ($tasks as $task) {
-            $updatePayload[] = [
-                'id' => $task->getId(),
-                'status' => ScheduledTaskDefinition::STATUS_QUEUED,
-            ];
-        }
+        $context = Context::createDefaultContext();
 
-        $this->scheduledTaskRepository->update($updatePayload, Context::createDefaultContext());
-
-        // Tasks **must not** be queued before their state in the database has been updated. Otherwise
+        // Tasks **must not** be queued before their state in the database has been updated. Otherwise,
         // a worker could have already fetched the task and set its state to running before it gets set to
         // queued, thus breaking the task.
         /** @var ScheduledTaskEntity $task */
         foreach ($tasks as $task) {
+            $this->scheduledTaskRepository->update([
```

Comment:
We should check and continue with next task when the task is already queued meanwhile.
This might happen, when queuing is running long f.e.

Reply:
Thinking about this, I reckon the core problem is scheduling tasks via a messenger service will never be 100% bullet proof against accidental multiple queuing. There just is no transaction that embraces the status update and the actual dispatch to the message queue.

Having said that, your concern appears reasonable to me, though academic. (And, btw., applies to the existing code as well.) 

The issue that I attempt to fix is a realistic scenario, namely some kind of runtime exception in the dispatch process. (Actually observed with client installations.)

On the other hand, in your scenario the assumption is that the dispatch process would take longer than ... what? Users would have to run scheduled-task:run in a frequency that is shorter than the presumed runtime for the dispatch process to cause such an issue. (How long does it take to dispatch a message? Typical runtime of scheduled-task:run is in the milli seconds, whilst "recommended" frequency of scheduled-task:run is 300 seconds).

So what would be your suggestion? To re-check the "scheduled" status of the task immediately prior to dispatch? I think we would just introduce an additional point of failure without gaining significant reliability, and thus would not recommend this.

I believe that, should a client require a 100% bullet proof solution for tasks not being dispatched multiply, the scheduled task itself would need to make use of a locking mechanism, like Symfony lock component.

Reply:
Couldn't we just use there a Lock to not have this issue? https://symfony.com/doc/current/components/lock.html

Reply:
I think we could, but locking would have to take place in the handle method of the MessageHandler, perhaps controlled by a property of the message.

Looks to me it would blow up the logic with little benefit. For "everyday usage" of Scheduled Tasks it might not be worth while. Users seem to have lived with this even with SW5, and I have seen locking implemented in custom tasks/cronjobs when really needed.

Reply:
What about locking the table using plain sql and setting the status then?

We don't need to use here DAL

Reply:
Sorry, I don't get it - what's the point to lock the scheduled_task table?

In the current code, when a task can't be queued because of an exception in the dispatch process, it will still show up as 'queued' in the table. Not nice, but at least prevents it from being unsuccessfully re-queued again and again. Using commit/rollback within a try/catch block, we could leave it in 'scheduled' - but that's not really better.

We would have to have a new state like 'queuing failed' to reflect that situation.